### PR TITLE
[`core`] Small refactor of forward pass

### PR DESCRIPTION
**Pull request by younesbelkada**
Let's address #133 step by step

This PR just refactors the`train_mini_batch` function, so that the loss function does not have to compute the logits internally - which makes the latest easier to test

Logs:

- gpt2-sentiment: https://wandb.ai/distill-bloom/trl/runs/j2v024qq?workspace=user-younesbelkada
- t5-sentiment: https://wandb.ai/distill-bloom/trl/runs/ncbwpxfx?workspace=user-younesbelkada

### CODE REVIEW

**Review by lvwerra — APPROVED** [review]

**Comment by younesbelkada** [comment]
Final runs: 
T5: https://wandb.ai/distill-bloom/trl/runs/7bxai1m5?workspace=user-younesbelkada
gpt2: https://wandb.ai/distill-bloom/trl/runs/2453mt4s?workspace=user-younesbelkada
